scripts/aws_replicate.py

```diff
@@ -0,0 +1,376 @@
+    
+    Returns:
+        None
+    """
+
+    config_chunk_size = global_config.get("chunk_size", 1)
+
+    index = 0
+
+    while index < len(files):
+        base_cmd = 'aws s3 cp s3://{} s3://{} --recursive --exclude "*"'.format(
+            source_bucket, target_bucket
+        )
+
+        chunk_size = min(config_chunk_size, len(files) - index)
+
+        execstr = base_cmd
+        for fi in files[index : index + chunk_size]:
+            object_name = "{}/{}".format(fi.get("id"), fi.get("filename"))
+
+            # only copy ones not exist in target bucket
+            if object_name not in copied_keys:
+                execstr += ' --include "{}"'.format(object_name)
+
+        if execstr != base_cmd:
+            subprocess.Popen(shlex.split(execstr + " --quiet")).wait()
+
+        index = index + chunk_size
+
+    global TOTAL_PROCESSED_FILES
+    mutexLock.acquire()
+    TOTAL_PROCESSED_FILES += len(files)
+    mutexLock.release()
+
+
+def process_data(threadName, source_bucket, global_config, copied_objects, q):
+    """
+    start to copy objects
+
+    Args:
+        threadName(str): thread name
+        source_bucket(str): source bucket
+        global_config(dict): configuration dictionary
+        copied_objects(set): objects aready copied
+        q(Queue): queue containing the objects need to be processed
+    """
+    while not EXIT_FLAG:
+        mutexLock.acquire()
+        if not q.empty():
+            files = q.get()
+            mutexLock.release()
+            target_bucket = get_bucket_name(files[0], PROJECT_MAP)
+            logger.info(
+                "%s processing %d to %s" % (threadName, len(files), target_bucket)
+            )
+            exec_aws_copy(
+                threadName,
+                files,
+                source_bucket,
+                target_bucket,
+                global_config,
+                copied_objects,
+            )
+        else:
+            mutexLock.release()
+
+
+def check_and_index_the_data(files, copied_objects, json_log):
+    """
+    batch list all the objects and check if file in manifest is copied or not.
+    Index data and log
+    """
+
+    for fi in files:
+        object_path = "{}/{}/{}".format(
+            get_bucket_name(fi, PROJECT_MAP), fi.get("id"), fi.get("filename")
+        )
+        logger.info("start to index {}".format(object_path))
+        if object_path not in copied_objects:
+            json_log[object_path] = {"copy_success": False, "index_success": False}
+        else:
+            try:
+                json_log[object_path] = {
+                    "copy_success": True,
+                    "index_success": update_indexd(fi),
+                }
+            except Exception as e:
+                logger.error(e.message)
+                json_log[object_path] = {
+                    "copy_success": True,
+                    "index_success": False,
+                    "msg": e.message,
+                }
+
+
+def update_indexd(fi):
```

Comment:
there's a lot of duplication between this and the google one, could these be consolidated?

Reply:
Fixed in d47a36c889171c98bbe27a87e89fb6d55d356f0f